packages/gatsby/src/commands/build.ts

```diff
@@ -201,6 +202,20 @@ module.exports = async function build(program: IBuildArgs): Promise<void> {
     buildSSRBundleActivityProgress.end()
   }
 
+  if (_CFLAGS_.GATSBY_MAJOR === `4` && shouldGenerateEngines()) {
+    const validateEnginesActivity = report.activityTimer(`Validating engines`, {
+      parentSpan: buildSpan,
+    })
+    validateEnginesActivity.start()
+    try {
+      await validateEngines(store.getState().program.directory)
+    } catch (err) {
+      validateEnginesActivity.panic(err)
```

Comment:
Let's use a structured error here with a new ID so that we can track it better in telemetry. We might also need to point people to _some_ docs/explanation as the error as is isn't actionable. Why is the import not allowed? How can people find the culprit (the "from" portion is only ever pointing to our query-engine, not the actual source?)? How can the fix it themselves?

Reply:
Yeah, agree - I did add this to TODO (in PR description), but before I do that, I think first we should assess wether this method of validation actually make sense conceptually (spawning short lived process and intercepting module loading there is way to go)

Reply:
I don't think we have many options 😉 At least I have no other ideas - maybe just copying (or symlinking) the engine to some temp folder. But that probably has its own set of pitfalls not to mention it won't fail on "require by absolute paths".

Reply:
Just going back on what error message should be here:

> We might also need to point people to some docs/explanation as the error as is isn't actionable.

I don't think this error ever will be really actionable to users, other than asking them to submit bug report or something along those lines. As it is right now users don't really have way to fix it on their own (other than maybe changing plugin that might cause the problem :yikes:)

> How can people find the culprit (the "from" portion is only ever pointing to our query-engine, not the actual source?)? How can they fix it themselves?

Right, so the error message as presented was just containing information that helped me assess that validation logic is correct - it was my intention to adjust message itself after checking with you about way validation works in this PR. Also error might point to `page-ssr` engine (not the query-engine one), so there is some value, but I'll see if I can get a hold of actual source code location (similar to how we print code frames for html rendering errors).

Reply:
> but I'll see if I can get a hold of actual source code location (similar to how we print code frames for html rendering errors).

Made it work, but it required generating source maps pretty much for engines and that seemed to slow down bundling quite a bit and I don't think tradeoff is worth it, so I just made error structured with actionability being "report a bug here"